mixins/textInput.js

```diff
@@ -114,29 +114,37 @@ export default {
     })
   },
 
+  iconClass() {
+    const props = this.props
+    const state = this.state
+    const hasServerErrors = (props.errors || []).length
+    return !props.showIcon ? false
+     : state.waiting ? 'refresh'
+     : (state.valid && !hasServerErrors) ? 'check'
+     : state.hasError ? 'times'
+     : props.disabled ? 'minus'
+     : props.icon ? props.icon
+     : (props.required && !props.value) ? 'caret-left'
+     : false
+  },
+
+  hasIcon() {
+    return !!this.iconClass()
+  },
+
   renderIcon() {
     let props = this.props
-    let errors = props.errors || []
-    let hasServerErrors = errors.length
     let className = classnames({
       'hui-TextInput__icon': true,
       'hui-TextInput__icon--left': (props.iconPosition === 'left')
     })
-    let state = this.state
-    let icon = !props.showIcon ? false
-               : state.waiting ? 'refresh'
-               : (state.valid && !hasServerErrors) ? 'check'
-               : state.hasError ? 'times'
-               : props.disabled ? 'minus'
-               : props.icon ? props.icon
-               : (props.required && !props.value) ? 'caret-left'
-               : false;
-
-    return icon && (
+    let icon = this.iconClass()
+
+    return this.hasIcon() && (
```

Comment:
Is this.hasIcon() necessary? you already have the variable **icon**, so this could read ```return !!icon && (```

seems it could be refactored to just a single method: this.icon(), since that would always return a truthy/falsy value which is also useful as the icon itself?